layerindex/tools/import_layers.py

```diff
@@ -0,0 +1,494 @@
+    rq = urllib.request.Request(branches_url)
+    data = urllib.request.urlopen(rq).read()
+    jsdata = json.loads(data.decode('utf-8'))
+    branch_idmap = {}
+    filter_branches = []
+    if args.branch:
+        for branch in args.branch.split(','):
+            if not Branch.objects.filter(name=branch).exists():
+                logger.error('"%s" is not a valid branch in this database (branches must be created manually first)' % branch)
+                sys.exit(1)
+            filter_branches.append(branch)
+    for branchjs in jsdata:
+        if filter_branches and branchjs['name'] not in filter_branches:
+            logger.debug('Skipping branch %s, not in specified branch list' % branchjs['name'])
+            continue
+        res = Branch.objects.filter(name=branchjs['name'])
+        if res:
+            branch = res.first()
+            branch_idmap[branchjs['id']] = branch
+        else:
+            logger.debug('Skipping branch %s, not in database' % branchjs['name'])
+
+    if args.layer:
+        layer_re = re.compile('^' + args.layer + '$')
+    else:
+        layer_re = None
+
+    try:
+        with transaction.atomic():
+            # Get layers
+            logger.info('Importing layers')
+            rq = urllib.request.Request(layers_url)
+            data = urllib.request.urlopen(rq).read()
+            jsdata = json.loads(data.decode('utf-8'), object_hook=datetime_hook)
+
+            layer_idmap = {}
+            exclude_fields = ['id', 'updated']
+            for layerjs in jsdata:
+                if layer_re and not layer_re.match(layerjs['name']):
+                    #logger.debug('Skipping layer %s, does not match layer restriction' % layerjs['name'])
+                    continue
+
+                layeritem = LayerItem.objects.filter(name=layerjs['name']).first()
+                if layeritem:
+                    # Already have this layer
+                    if layerjs['updated'] <= layeritem.updated and not args.reload:
+                        logger.debug('Skipping layer %s, already up-to-date' % layerjs['name'])
+                        layer_idmap[layerjs['id']] = layeritem
+                        continue
+                    else:
+                        logger.debug('Updating layer %s' % layerjs['name'])
+                else:
+                    logger.debug('Adding layer %s' % layerjs['name'])
+                    layeritem = LayerItem()
+                for key, value in layerjs.items():
+                    if key in exclude_fields:
+                        continue
+                    setattr(layeritem, key, value)
+                layeritem.save()
+                layer_idmap[layerjs['id']] = layeritem
+
+            # Get layer branches
+            logger.debug('Importing layer branches')
+            rq = urllib.request.Request(layerbranches_url)
+            data = urllib.request.urlopen(rq).read()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *urlopen*. The resource is allocated by call *request.urlopen*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close the object returned by *urlopen()* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).